core/domain/stats_jobs_one_off.py

```diff
@@ -195,3 +196,248 @@ def reduce(key, stringified_values):
+                _ANSWER_BUCKET_REMOVED_KEY with an interaction ID appended.
+            stringified_values: list(str). A list of stringified values
+                associated with the given key. An element of stringified_values
+                would be of the form:
+                    {
+                        'item_id': str. the id of the mapped model.
+                        'interaction_id': str. the corresponding interaction ID.
+                    }
+
+        Yields:
+            tuple(str). A 1-tuple whose only element is an output message.
+        """
+        if key.startswith(PurgeNewAnswersAuditJob._ANSWER_BUCKET_REMOVED_KEY):
+            first_value_dict = ast.literal_eval(stringified_values[0])
+            interaction_id = first_value_dict['interaction_id']
+            item_ids = [
+                ast.literal_eval(stringified_value)['item_id']
+                for stringified_value in stringified_values]
+            yield (
+                'Will remove %d answer buckets corresponding to purged '
+                'interaction %s: %s' % (
+                    len(stringified_values), interaction_id, item_ids))
+        else:
+            yield (
+                'Will retain %d answer buckets corresponding to current '
+                'interactions' % len(stringified_values))
+
+
+class PurgeNewAnswersJob(jobs.BaseMapReduceJobManager):
+    """A one-off purge job for answers.
+
+    Purges all answers from StateAnswersModel which correspond to interactions
+    that are no longer being used by Oppia, such as FileReadInput and
+    TarFileReadInput. See #3284 for additional context.
+    """
+    _ANSWER_BUCKET_REMOVED_KEY = 'removed_answer_bucket'
+    _ANSWER_BUCKET_RETAINED_KEY = 'retained_answer_bucket'
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [stats_models.StateAnswersModel]
+
+    @staticmethod
+    def map(item):
+        """Implements the map function and removes all answer buckets associated
+        with interactions that no longer exist.
+
+        Args:
+            item: StateAnswersModel.
+
+        Yields:
+            tuple. a 2-tuple in the form
+                (output_key, value).
+                'output_key': str. either _ANSWER_BUCKET_RETAINED_KEY or
+                    _ANSWER_BUCKET_REMOVED_KEY with an interaction ID appended.
+                'value': a dict, whose structure is as follows:
+                    {
+                        'item_id': str. the id of the mapped model.
+                        'interaction_id': str. the corresponding interaction ID.
+                    }
+        """
+        interaction_ids = (
+            interaction_registry.Registry.get_all_interaction_ids())
+        if item.interaction_id not in interaction_ids:
+            item.delete()
+            removal_key = '%s-%s' % (
+                PurgeNewAnswersJob._ANSWER_BUCKET_REMOVED_KEY,
+                item.interaction_id)
```

Comment:
We have had jobs in the past delete items in the mapper. I think it's fine since mapreduce already has a collection to iterate across. There's also some precedence found online for using deletion in the map step.

I would expect to be able to still access ``item.id`` and ``item.interaction_id`` after deletion, since deletion should be an RPC and not affect the local snapshot of the data. I copied the values to local variables, anyway, to be safe about it.